src/goto-instrument/contracts/contracts.cpp

```diff
@@ -1260,3 +1259,34 @@ bool code_contractst::enforce_contracts(const std::set<std::string> &to_enforce)
   }
   return fail;
 }
+
+void code_contractst::find_static_locals(
+  const goto_functionst &functions,
+  const irep_idt &root_function,
+  assigns_clauset &assigns)
+{
+  auto cg =
+    call_grapht::create_from_root_function(functions, root_function, true);
+  auto g = cg.get_directed_graph();
+
+  // Adds local static declarations to write set
+  for(const auto &sym_pair : symbol_table)
+  {
+    const auto &sym = sym_pair.second;
+    if(sym.is_static_lifetime)
+    {
+      auto fname = sym.location.get_function();
+      if(
+        !fname.empty() &&
+        (fname == root_function || g.get_node_index(fname).has_value()))
+      {
+        // the symbol has
+        // - a static lifetime
+        // - non empty location function
+        // - this function appears in the call graph of the function
+        assigns.add_to_write_set(sym.symbol_expr());
```

Comment:
What about function pointers?

Reply:
we suppose funciton pointers have been removed/restricted before this instrumentation phase so that the target would appear as plain calls in the program

Reply:
Hmm, we should really include a `PRECONDITION` in `call_graph.cpp`'s `forall_callsites` to replace `if(function_expr.id()==ID_symbol)` by `PRECONDITION_WITH_DIAGNOSTICS(function_expr.id() == ID_symbol, "call graph computation requires function pointer removal");`.

Reply:
`PRECONDITION` added in `call_graph.cpp`. 


Should we also modify `goto_function_inline` to warn on unresolved function pointers ?

Reply:
Yes, we should either warn or fail with a `PRECONDITION`.